ni.zsh

```diff
@@ -571,18 +571,88 @@ function ni-dlx(){
 
 # auto completion
 function _ni(){
-  # ni <subcommands>
-  local -a subcommands
-  subcommands=(
-    'add:add package'
-    'run:run scripts'
-    'test:run test script'
-    'upgrade:upgrade package'
-    'upgrade-interactive:upgrade package interactively'
-    'remove:remove package'
-    'exec:execute command'
-    'dlx:download package and execute command'
-  )
-  _describe -t subcommands 'subcommands' subcommands
+  local context state state_descr line
+  typeset -A opt_args
+
+  _arguments -C \
+    '1: :->cmds' \
+    '*:: :->args'
+
+  case "$state" in
+    cmds)
+      # ni <subcommands>
+      local -a subcommands
+      subcommands=(
+        'add:add package'
+        'run:run scripts'
+        'test:run test script'
+        'upgrade:upgrade package'
+        'upgrade-interactive:upgrade package interactively'
+        'remove:remove package'
+        'exec:execute command'
+        'dlx:download package and execute command'
+      )
+      _describe -t subcommands 'subcommands' subcommands
+      ;;
+    args)
+      case $line[1] in
+        add)
+          # ni add <package>
+          _arguments \
+            '--dev[Install as development dependency]' \
+            '-D[Install as development dependency]'
+          ;;
+        remove|upgrade)
+          # ni remove <package>
+          # ni upgrade <package>
+          if [ -f "package.json" ]; then
+            if command -v jq >/dev/null 2>&1; then
+              local -a packages
+              packages=(${(f)"$(cat package.json | jq -r '.dependencies // {} | to_entries | .[] | select(.key != "") | "\(.key):\(.value)"')"})
```

Comment:
What does `${(f)` mean?

Reply:
There might not be a common name.  
It is sometimes referred to as a line-splitting flag. 
Ref: https://gist.github.com/sho-t/d9cdf8271b3de7c4238739e523490542#triangular_flag_on_post-%E5%A4%89%E6%95%B0%E5%B1%95%E9%96%8B%E3%83%95%E3%83%A9%E3%82%B0parameter-expansion-flags-zzz